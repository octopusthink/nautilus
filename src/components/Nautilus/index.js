import { Global, css } from '@emotion/core';
import { ThemeProvider } from 'emotion-theming';
import PropTypes from 'prop-types';
import React from 'react';

import { nautilus as nautilusDefaultTheme } from 'themes';

export const Nautilus = ({ children, theme }) => {
  return (
    <ThemeProvider theme={theme}>
      <Global
        styles={css`
          html {
            /* Makes type rendering look crisper. */
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;

            /* Set base font size for rem units to inherit .*/
            font-size: 62.5%;

            /* Enable kerning and optional ligatures. */
            text-rendering: optimizeLegibility;

            /* Disable auto-enlargement of small text in Safari. */
            text-size-adjust: 100%;
          }
        `}
      />
      {children}
    </ThemeProvider>
  );
};

Nautilus.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

Nautilus.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Theme object used to style this instance of Nautilus and its components. */
  theme: PropTypes.object,
};

export default Nautilus;
