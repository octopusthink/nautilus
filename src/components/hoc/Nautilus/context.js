import { createContext } from 'react';

export const NautilusLinkComponentContext = createContext();
