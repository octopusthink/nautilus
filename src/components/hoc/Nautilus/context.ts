import { ReactElement, createContext } from 'react';

import LinkComponent from '../../ui/Link';

export const NautilusLinkComponentContext = createContext<ReactElement | any>(LinkComponent);
