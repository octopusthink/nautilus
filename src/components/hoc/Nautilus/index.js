import { Global, css } from '@emotion/core';
import { ThemeProvider } from 'emotion-theming';
import PropTypes from 'prop-types';
import React from 'react';

import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

export const Nautilus = ({ children, theme }) => {
  return (
    <ThemeProvider theme={theme}>
      <Global
        styles={css`
          * {
            /* For the love of everything, use better box-sizing by default. */
            box-sizing: border-box;
          }

          html {
            /* Set base font size for rem units to inherit. */
            font-size: 62.5%;

            /* Makes type rendering look crisper. */
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;

            /* Enable kerning and optional ligatures. */
            text-rendering: optimizeLegibility;

            /* Disable auto-enlargement of small text in Safari. */
            text-size-adjust: 100%;
          }
        `}
      />
      {children}
    </ThemeProvider>
  );
};

Nautilus.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

Nautilus.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Theme object used to style this instance of Nautilus and its components. */
  theme: themePropTypes,
};

export default Nautilus;
