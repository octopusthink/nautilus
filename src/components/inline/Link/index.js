import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';
import { Link as ReactRouterLink } from 'react-router-dom';

export const Link = (props) => {
  const { children, ...otherProps } = props;

  return <ReactRouterLink {...otherProps}>{children}</ReactRouterLink>;
};

export const styles = () => {
  return css``;
};

Link.defaultProps = {
  children: undefined,
};

Link.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

export const { defaultProps, propTypes } = Link;

export default styled(Link)(styles);
