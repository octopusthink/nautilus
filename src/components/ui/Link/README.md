A `<Link>` provides a way of navigating to other content, either on the same site or elsewhere on the web.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Paragraph>
  Fill out the <Link href="#sign-up">sign up form</Link> to create your account.
</Paragraph>
```

## Usage

Use a `<Link>` component where you would usually use an `<a>` tag in HTML, including for navigation within your application.

### Use this component for…

- Navigating to other pages or content, on the same site or another site.
- Creating an inline link in a paragraph or other span of text.
- Creating a link to wrap around multiple elements (eg: an image and some text).

### Don't use this component for…

- Primary navigation for your site. Use a `<Menu>` instead.
- Performing an action. Use a `<Button>` instead.
- A standalone navigational link. Use a `<Button navigation>` instead.

## Appearance

Links are styled consistently throughout, using a distinct colour and a slightly lighter underline. Colours are adjustable in your theme.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Paragraph>
  Looking for a new furry friend? Browse our
  <Link href="#puppies">puppies</Link> and
  <Link href="#kittens">kittens</Link> for adoption.
</Paragraph>;
```

## Behaviour

### Links to external resources

Use the `external` prop to indicate that a link navigates to an off-site, external resource. This helps users know what to expect before they click a link.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Paragraph>
  Make sure to read the
  <Link href="https://tpsreports.com" external>TPS Reports</Link> prior
  to the Monday Mega-Meeting.
</Paragraph>
```

### Using a router

By default, Nautilus will output an `<a>` tag for every `Link` component.

If you're using a router (like React Router or Reach Router), you'll probably want to use the router's `Link` component under the hood.

To do this, set a config value for `LinkComponent` in your [`Nautilus`](/#/Function/Higher-order%20components/Nautilus) higher-order component.

Here's how you would use various link components from popular router libraries:

```jsx
import Nautilus, { Paragraph, Tabs } from '@octopusthink/nautilus';
import { Link as ReachRouterLink } from '@reach/router';
import { createMemoryHistory } from 'history';
import { Link as ReactRouterLink, Router } from 'react-router-dom';

<Tabs>
  <Tabs.Tab label="Reach Router">
    <Nautilus config={{ LinkComponent: ReachRouterLink }}>
      <Paragraph>
        This link uses <Link to="#react-router">Reach Router</Link>.
      </Paragraph>
    </Nautilus>
  </Tabs.Tab>

  <Tabs.Tab label="React Router">
    <Nautilus config={{ LinkComponent: ReactRouterLink }}>
      <Router history={createMemoryHistory()}>
        <Paragraph>
          But this link uses <Link to="#react-router">React Router</Link>.
        </Paragraph>
      </Router>
    </Nautilus>
  </Tabs.Tab>
</Tabs>
```

### Overriding individual Link components

If you want to override a specific `Link` component, you can use the `as` prop:

```jsx
import Nautilus, { Paragraph } from '@octopusthink/nautilus';
import { createMemoryHistory } from 'history';
import { Link as ReactRouterLink, Router } from 'react-router-dom';

<Nautilus>
  <Router history={createMemoryHistory()}>
    <Paragraph>
      Fill out the
      <Link as={ReactRouterLink} to="/sign-up">sign up form</Link> to
      create your account.
    </Paragraph>
  </Router>
</Nautilus>
```

This is especially useful if you have set a default `Link` component that you want to use for the majority of links, but occassionally want to output a regular `<a>` tag.

## Voice & tone

- Link text should be short and succinct. Aim for four words or less.
- Link text should clearly label the intended destination. Use the destination title where possible.
- Don't use "click here" or another link to "here".
- Don't use the web address itself, unless it's the most meaningful descriptive label for the link.

## Accessibility

### Use a clear, unique text label for link text

Use text for links. If using graphics or icons, provide a text alternative or pair them with a text label.

Since screen reader users may choose to skim all the links in a page without reading the surrounding content, it's especially important to ensure that your link text works on its own and accurately labels the intended destination. Test your link text to make sure it's sufficiently descriptive and unique when read on its own. Don't use "here" as your link text.

If you have multiple links on a page that navigate to similar, but unique, content, make sure they have unique labels. For instance, if you're showing blog post previews followed by a "Read more" link, ensure the title of the post is somewhere in the link so that screen reader users can differentiate the posts.

### Don't open links in new windows

Don't open links in new windows. This may be done with good intent, but it often has a negative impact on the usability and accessibility of the page and can disorient users.

Opening links in new windows hijacks the user experience and forces users to open new tabs, whether they want to or not. Instead, use an `external` indicator to show that the link leads to off-site content, and allow users to choose whether they want that content to open in a new tab or not.

### Ensure links are clearly identified

It's best to use two different indicators for links: an underline and an accent colour. This is a long-established pattern for identifying links on the web, so be extremely careful deviating for this pattern.

Never rely on colour or hover states alone to differentiate links. Be very careful with underlining text that isn't a link, since users may expect these to function as links.
