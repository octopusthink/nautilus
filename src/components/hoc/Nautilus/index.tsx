import { Global, css } from '@emotion/core';
import { ThemeProvider } from 'emotion-theming';
import React, { ReactNode } from 'react';

import { nautilusDefaultTheme } from '../../../themes';
import { NautilusLinkComponentContext } from './context';

interface Props {
  /** @ignore */
  children?: React.ReactNode;
  /** Config values affect all component rendered underneath this `Nautilus` component. Used to set default `Link` components, etc. */
  config?: {
    LinkComponent?: ReactNode;
  };
  /** Theme object used to style this instance of Nautilus and its components. */
  theme?: Theme;
}

const Nautilus = ({
  children,
  config = { LinkComponent: undefined },
  theme = nautilusDefaultTheme,
}: Props) => {
  const { LinkComponent } = config || {};

  return (
    <NautilusLinkComponentContext.Provider value={LinkComponent}>
      <ThemeProvider theme={theme}>
        <Global
          styles={css`
            * {
              /* For the love of everything, use better box-sizing by default. */
              box-sizing: border-box;
            }

            html {
              /* Set base font size for rem units to inherit. */
              font-size: 62.5%;

              /* Makes type rendering look crisper. */
              -moz-osx-font-smoothing: grayscale;
              -webkit-font-smoothing: antialiased;

              /* Enable kerning and optional ligatures. */
              text-rendering: optimizeLegibility;

              /* Disable auto-enlargement of small text in Safari. */
              text-size-adjust: 100%;
            }
          `}
        />
        {children}
      </ThemeProvider>
    </NautilusLinkComponentContext.Provider>
  );
};

export default Nautilus;
