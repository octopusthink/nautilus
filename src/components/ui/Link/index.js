import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React, { useContext } from 'react';

import { NautilusLinkComponentContext } from '../../hoc/Nautilus/context';
import Icon from '../Icon';
import { focusStyle, toUnits } from '../../../styles';
import { useTheme } from '../../../themes';

const LinkTag = 'a';

const Link = (props) => {
  const defaultLinkComponent = useContext(NautilusLinkComponentContext);
  const theme = useTheme();

  const { children, as, external, href, unstyled, ...otherProps } = props;
  const LinkComponent = as || (href && 'a') || (external && 'a') || defaultLinkComponent || LinkTag;

  return (
    <LinkComponent
      css={
        unstyled
          ? undefined
          : css`
              color: ${theme.colors.state.interactiveText};
              box-shadow: 0 2px ${theme.colors.state.interactive};
              text-decoration: none;
              transition: all 200ms ease-in-out;

              &:hover {
                box-shadow: 0 2px ${theme.colors.state.hover};
                color: ${theme.colors.state.hoverText};
              }

              &:focus {
                ${focusStyle.outlineLight(theme)};
                ${focusStyle.text(theme)};
              }
            `
      }
      href={href}
      {...otherProps}
    >
      {children}
      {external && (
        <React.Fragment>
          <Icon
            name="external-link"
            noMargin
            small
            strokeColor={theme.colors.neutral.grey600}
            title="External link"
            css={css`
              margin-left: ${toUnits(theme.spacing.padding.xSmall)};
            `}
          />
        </React.Fragment>
      )}
    </LinkComponent>
  );
};

Link.defaultProps = {
  as: undefined,
  children: undefined,
  external: false,
  href: undefined,
  unstyled: false,
};

Link.propTypes = {
  /** Component/tag to render the underlying link. Defaults to `Nautilus.config.defaultComponents.Link` if set; an `<a>` tag will be used if `Nautilus.config.defaultComponents.Link` is not set. */
  as: PropTypes.elementType,
  /** @ignore */
  children: PropTypes.node,
  /** Use to indicate that a link points to an external resource. */
  external: PropTypes.bool,
  /** The URL to navigate to when this link is activated. */
  href: PropTypes.string,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Link;

export default Link;
