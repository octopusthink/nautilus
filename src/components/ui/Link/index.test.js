import { Link as ReachRouterLink } from '@reach/router';
import React, { createRef } from 'react';

import { axe, render } from 'utils/testing';

import Link from '.';
import Nautilus from '../../hoc/Nautilus';

describe('Link', () => {
  it('should render an <a> tag', () => {
    const { container } = render(<Link href="/nowhere/">Hello</Link>);

    expect(container.firstChild.tagName).toEqual('A');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Link href="/nowhere/">
        <span data-testid="child" />
      </Link>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Link className="custom-class" data-testid="myText" href="/nowhere/">
        hello
      </Link>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should output an href prop', () => {
    const { container } = render(<Link href="/a-url/">Hello</Link>);

    expect(container.firstChild.getAttribute('href')).toEqual('/a-url/');
  });

  it('should add an icon when `external` is set', () => {
    const { container } = render(
      <Link href="https://elsewhere.net/" external>
        hello
      </Link>,
    );

    expect(container.firstChild.tagName).toEqual('A');
    expect(container.firstChild.children[0].children[0].tagName).toEqual('svg');
  });

  it('should output Link styles', () => {
    const { container } = render(
      <Link href="https://elsewhere.net/">hello</Link>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should forward refs ', () => {
    const ref = createRef();

    render(
      <Link href="/" ref={ref}>
        Homepage
      </Link>,
    );

    expect(ref.current).not.toBeNull();
    expect(ref.current.tagName).toEqual('A');
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<Link href="/">Homepage</Link>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });

  describe('with default Link components', () => {
    it('should output an href prop when the to prop is supplied', () => {
      const { container } = render(
        <Nautilus LinkComponent={ReachRouterLink}>
          <Link to="/a-url/">Hello</Link>
        </Nautilus>,
      );

      expect(container.firstChild.getAttribute('href')).toEqual('/a-url/');
    });
  });

  describe('with custom Link components', () => {
    it('should output an href prop when the to prop is supplied', () => {
      const { container } = render(
        <Link asComponent={ReachRouterLink} to="/a-url/">
          Hello
        </Link>,
      );

      expect(container.firstChild.getAttribute('href')).toEqual('/a-url/');
    });
  });
});
