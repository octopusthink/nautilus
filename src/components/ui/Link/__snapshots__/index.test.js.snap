// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should not output styles when __unstyled is true 1`] = `
<div>
  <a
    class="emotion-0"
    href="https://octopusthink.com/"
  >
    hello
  </a>
</div>
`;

exports[`Link should output Link styles 1`] = `
.emotion-0 {
  border-bottom: 2px solid #f5bbda;
  color: #cd2f83;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 200ms ease-in-out;
  transition: all 200ms ease-in-out;
}

.emotion-0:hover {
  border-color: #a0dee3;
  color: #0097a2;
}

.emotion-0:focus {
  border-color: transparent;
  box-shadow: #f5bbda 0px 0px 0px 3px;
  outline: none;
}

<div>
  <a
    class="emotion-0"
    href="https://octopusthink.com/"
  >
    hello
  </a>
</div>
`;
